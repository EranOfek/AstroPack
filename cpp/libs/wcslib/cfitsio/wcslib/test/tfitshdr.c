/*============================================================================
  WCSLIB 7.3 - an implementation of the FITS WCS standard.
  Copyright (C) 1995-2020, Mark Calabretta

  This file is part of WCSLIB.

  WCSLIB is free software: you can redistribute it and/or modify it under the
  terms of the GNU Lesser General Public License as published by the Free
  Software Foundation, either version 3 of the License, or (at your option)
  any later version.

  WCSLIB is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
  more details.

  You should have received a copy of the GNU Lesser General Public License
  along with WCSLIB.  If not, see http://www.gnu.org/licenses.

  Direct correspondence concerning WCSLIB to mark@calabretta.id.au

  Author: Mark Calabretta, Australia Telescope National Facility, CSIRO.
  http://www.atnf.csiro.au/people/Mark.Calabretta
  $Id: tfitshdr.c,v 7.3.1.2 2020/08/17 11:20:48 mcalabre Exp mcalabre $
*=============================================================================
*
* tfitshdr tests fitshdr(), the FITS parser for image headers, by reading a
* test header and printing the resulting fitskey structs.
*
* Input comes from file 'fitshdr.fits' using either fits_hdr2str() from
* CFITSIO if the DO_CFITSIO preprocessor is defined, or read directly using
* fgets() otherwise.
*
* wcshdr() is called first to extract all WCS-related keyrecords from the
* input header before passing it on to fitshdr().
*
*---------------------------------------------------------------------------*/

#include <wcsconfig_tests.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#if defined HAVE_CFITSIO && defined DO_CFITSIO
#include <fitsio.h>
#endif

#include <fitshdr.h>
#include <wcshdr.h>
#include <wcsutil.h>

int main()

{
  char infile[] = "fitshdr.fits";
  char text[80];
  int  i, j, k, nkeyrec, nkeyids, nreject, status;
  struct fitskey *keys, *kptr;
  struct fitskeyid keyids[8];
#if defined HAVE_CFITSIO && defined DO_CFITSIO
  char *header;
  fitsfile *fptr;
#else
  char keyrec[81], header[288001];
  int  end;
  FILE *fptr;
#endif

  struct wcsprm *wcs;
  int  ctrl, nwcs, relax;


  // Set line buffering in case stdout is redirected to a file, otherwise
  // stdout and stderr messages will be jumbled (stderr is unbuffered).
  setvbuf(stdout, NULL, _IOLBF, 0);

  printf("Testing FITS image header parser (tfitshdr.c)\n"
         "---------------------------------------------\n\n");

  // Read in the FITS header.
#if defined HAVE_CFITSIO && defined DO_CFITSIO
  status = 0;

  if (fits_open_file(&fptr, infile, READONLY, &status)) {
    fits_report_error(stderr, status);
    return 1;
  }

  if (fits_hdr2str(fptr, 0, NULL, 0, &header, &nkeyrec, &status)) {
    fits_report_error(stderr, status);
    return 1;
  }

  fits_close_file(fptr, &status);
#else
  if ((fptr = fopen(infile, "r")) == 0) {
    printf("ERROR opening %s\n", infile);
    return 1;
  }

  k = 0;
  end = 0;
  nkeyrec = 0;
  for (j = 0; j < 100; j++) {
    for (i = 0; i < 36; i++) {
      if (fgets(keyrec, 81, fptr) == 0) {
        break;
      }

      memcpy(header+k, keyrec, 80);
      k += 80;
      nkeyrec++;

      if (strncmp(keyrec, "END       ", 10) == 0) {
        // An END keyrecord was read, but read the rest of the block.
        end = 1;
      }
    }

    if (end) break;
  }
  fclose(fptr);
#endif

  printf("Found %d header keyrecords.\n", nkeyrec);


  // Cull recognized, syntactically valid WCS keyrecords from the header.
  relax = WCSHDR_all;
  ctrl = -1;
  if ((status = wcspih(header, nkeyrec, relax, ctrl, &nreject, &nwcs,
                       &wcs))) {
    fprintf(stderr, "wcspih ERROR %d: %s.\n", status, wcs_errmsg[status]);
    return 1;
  }
  wcsvfree(&nwcs, &wcs);

  // Number remaining.
  nkeyrec = strlen(header) / 80;


  // Specific keywords to be located or culled.
  strcpy(keyids[0].name, "SIMPLE  ");
  strcpy(keyids[1].name, "BITPIX  ");
  strcpy(keyids[2].name, "NAXIS   ");
  strcpy(keyids[3].name, "COMMENT ");
  strcpy(keyids[4].name, "HISTORY ");
  strcpy(keyids[5].name, "        ");
  strcpy(keyids[6].name, "END     ");
  nkeyids = 7;

  if (nkeyids) {
    printf("\nThe following keyrecords will not be listed:\n");
    for (i = 0; i < nkeyids; i++) {
      printf("  \"%8s\"\n", keyids[i].name);
    }
  }


  // Parse the header.
  if ((status = fitshdr(header, nkeyrec, nkeyids, keyids, &nreject, &keys))) {
    printf("fitskey ERROR %d: %s.\n", status, fitshdr_errmsg[status]);
  }
#if defined HAVE_CFITSIO && defined DO_CFITSIO
  fits_free_memory(header, &status);
#endif

  // Report the results.
  printf("\n%d header keyrecords parsed by fitshdr(), %d rejected:\n\n",
    nkeyrec, nreject);
  kptr = keys;
  for (i = 0; i < nkeyrec; i++, kptr++) {
    // Skip syntactically valid keyrecords that were indexed.
    if (kptr->keyno < 0 && !kptr->status) continue;

    // Basic keyrecord info.
    printf("%4d%5d  %-8s%3d", kptr->keyno, kptr->status, kptr->keyword,
                              kptr->type);

    // Format the keyvalue for output.
    switch (abs(kptr->type)%10) {
    case 1:
      // Logical.
      sprintf(text, "%c", kptr->keyvalue.i?'T':'F');
      break;
    case 2:
      // 32-bit signed integer.
      sprintf(text, "%d", kptr->keyvalue.i);
      break;
    case 3:
      // 64-bit signed integer.
#ifdef WCSLIB_INT64
         sprintf(text, "%+lld", kptr->keyvalue.k);
#else
         if (kptr->keyvalue.k[2]) {
           sprintf(text, "%+d%09d%09d", kptr->keyvalue.k[2],
                                    abs(kptr->keyvalue.k[1]),
                                    abs(kptr->keyvalue.k[0]));
         } else {
           sprintf(text, "%+d%09d",     kptr->keyvalue.k[1],
                                    abs(kptr->keyvalue.k[0]));
         }
#endif
       break;
    case 4:
      // Very long integer.
      k = 0;
      for (j = 7; j > 0; j--) {
        if (kptr->keyvalue.l[j]) {
          k = j;
          break;
        }
      }

      sprintf(text, "%+d", kptr->keyvalue.l[k]);
      for (j = k-1; j >= 0; j--) {
        sprintf(text+strlen(text), "%09d", abs(kptr->keyvalue.l[j]));
      }

      break;
    case 5:
      // Float.
      sprintf(text, "%+13.6e", kptr->keyvalue.f);
      break;
    case 6:
      // Int complex.
      sprintf(text, "%.0f  %.0f", kptr->keyvalue.c[0],
                                  kptr->keyvalue.c[1]);
      break;
    case 7:
      // Float complex.
      sprintf(text, "%+13.6e  %+13.6e", kptr->keyvalue.c[0],
                                        kptr->keyvalue.c[1]);
      break;
    case 8:
      // String.
      sprintf(text, "\"%s\"", kptr->keyvalue.s);
      break;
    default:
      // No value.
      *text = '\0';
      break;
    }

    if (kptr->type > 0) {
      // Keyvalue successfully extracted.
      printf("  %s", text);
    } else if (kptr->type < 0) {
      // Syntax error of some type while extracting the keyvalue.
      printf("  (%s)", text);
    }

    // Units?
    if (kptr->ulen) {
      printf(" %.*s", kptr->ulen-2, kptr->comment+1);
    }

    // Comment text or reject keyrecord.
    printf("\n%s\n", kptr->comment);
  }


  // Print indexes.
  printf("\n\nIndexes of selected keywords:\n");
  for (i = 0; i < nkeyids; i++) {
    printf("%-8s%5d%5d%5d", keyids[i].name, keyids[i].count, keyids[i].idx[0],
      keyids[i].idx[1]);

    // Print logical (SIMPLE) and integer (BITPIX, NAXIS) values.
    if (keyids[i].count) {
      kptr = keys + keyids[i].idx[0];
      printf("%4d", kptr->type);

      if (kptr->type == 1) {
        printf("%5c", kptr->keyvalue.i?'T':'F');
      } else if (kptr->type == 2) {
        printf("%5d", kptr->keyvalue.i);
      }
    }
    printf("\n");
  }

  wcsdealloc(keys);

  return 0;
}
