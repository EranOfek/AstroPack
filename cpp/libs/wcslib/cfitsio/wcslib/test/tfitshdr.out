Testing FITS image header parser (tfitshdr.c)
---------------------------------------------

Found 144 header keyrecords.

The following keyrecords will not be listed:
  "SIMPLE  "
  "BITPIX  "
  "NAXIS   "
  "COMMENT "
  "HISTORY "
  "        "
  "END     "

74 header keyrecords parsed by fitshdr(), 21 rejected:

   9    0  UNDEF     0
Undefined keyvalue
  10    0  TRUE      1  T
Logical
  11    0  FALSE     1  F
Logical
  12    0  INT32     2  12345
Not a 64-bit integer
  13    0  INT32     2  -123456789
Not a 64-bit integer
  14    0  INT32     2  -2147483648
Not a 64-bit integer (INT_MIN)
  15    0  INT32     2  2147483647
Not a 64-bit integer (INT_MAX)
  16    0  INT32     2  12345
Not a very long integer
  17    0  INT32     2  -123456789
Not a very long integer
  18    0  INT64     3  -2147483649
64-bit integer (INT_MIN - 1)
  19    0  INT64     3  +2147483648
64-bit integer (INT_MAX + 1)
  20    0  INT64     3  +100000000000000000
64-bit integer
  21    0  INT64     3  -876543210987654321
64-bit integer
  22    0  INT64     3  -9223372036854775808
Not a very long integer (LONG_MIN)
  23    0  INT64     3  +9223372036854775807
Not a very long integer (LONG_MAX)
  24    0  INT64     3  -876543210987654321
64-bit integer
  25    0  INTVL     4  -9223372036854775809
Very long integer (LONG_MIN - 1)
  26    0  INTVL     4  +9223372036854775808
Very long integer (LONG_MAX + 1)
  27    0  INTVL     4  -100000000000000000000000000000876543210987654321
Very-long integer
  28    0  INTVL     4  +123456789012345678901234567890123456789012345678901234567890123456789

  29    0  INTVL     4  +1234567890123456789012345678901234567890123456789012345678901234567890

  30    0  FLOAT     5  +3.141593e+00
Floating point
  31    0  FLOAT     5  +1.602177e-19
Floating point, lower-case exp allowed
  32    0  FLOAT     5  +2.997925e+08
Floating point
  33    0  FLOAT     5  +6.626070e+00
Floating point, lower-case exp allowed
  34    0  FLOAT     5  +6.022141e+00
Floating point
  35    0  COMPLEX   6  137  -1
An integer complex keyvalue
  36    0  COMPLEX   7  +1.000000e+06  -1.000000e-01
A floating point complex keyvalue
  37    0  GOODSTR   8  ""G'DAY""
A valid string keyvalue
  38    0  BLANKS    8  " "
An all-blank string equals a single blank
  39    0  LONGSTR   8  "The loooooongest possible non-continued string value, 68 characters."

  40    0  CONTSTR  18  "The quick brown "
Continued string, with & (will be stripped off)
  41    4  CONTINUE 28  "fox jumps over the"
CONTINUE  'fox jumps over the'   Relaxed handling of invalid keycomment         
  42    0  CONTINUE 38  " lazy dog"

  43    0  PI        5  +3.141593e+00
[!?] is not a valid units specification
  44    0  CONTINUE  0
      'Not a valid string continuation'
  50    1  BADREC =  0
                     1 / Illegal, "= " not in columns 9-10
  51    0  BADREC    0
 =                   1 / Illegal, "= " not in columns 9-10
  52    0  BADREC    0
=1                     / Illegal, "= " not in columns 9-10
  53    0  BADREC    0
                     1 / Illegal, "= " not in columns 9-10
  54    1  BadKey    2  111
Example of a syntax error, invalid keyword
  55    1  BADKEY!   2  222
Example of a syntax error, invalid keyword
  56    1  BAD KEY   2  333
Example of a syntax error, invalid keyword
  57    2  BADVAL    0
BADVAL  =                  BAD / Example of a syntax error, invalid keyvalue    
  58    2  BADVAL    0
BADVAL  =                "BAD" / Example of a syntax error, invalid keyvalue    
  59    4  BADVAL   -8  ("BAD")
BADVAL  =          'BAD' 'STR' / Example of a syntax error, invalid keyvalue    
  60    4  BADVAL   -2  (49)
BADVAL  =              49 + 94 / Example of a syntax error, invalid keyvalue    
  61    4  BADVAL    0
BADVAL  =             1934-638 / Example of a syntax error, invalid keyvalue    
  62    4  BADVAL    0
BADVAL  =                1.2.3 / Example of a syntax error, invalid keyvalue    
  63    4  BADVAL    0
BADVAL  =            12h34m56s / Example of a syntax error, invalid keyvalue    
  64    4  BADVAL    0
BADVAL  =                 0xFF / Example of a syntax error, invalid keyvalue    
  65    4  BADVAL   -2  (58)
BADVAL  =              58 - 21 / Example of a syntax error, invalid keyvalue    
  66    2  BADVAL    0
BADVAL  =                J2000 / Example of a syntax error, invalid keyvalue    
  67    2  BADVAL    0
BADVAL  =              B1950.0 / Example of a syntax error, invalid keyvalue    
  68    4  BADVAL    0
BADVAL  =            180:00:00 / Example of a syntax error, invalid keyvalue    
  69    2  BADVAL    0
BADVAL  =                $9.99 / Example of a syntax error, invalid keyvalue    
  70    4  BADCOMM  -2  (1)
BADCOMM =                    1   Example of a syntax error, invalid keycomment  
  71    4  BADCOMM  -2  (1)
BADCOMM =                    1   1/8 = 0.125                                    
 -73    8  END       8  "ILLEGAL"
Illegal END keyrecord


Indexes of selected keywords:
SIMPLE      1    0   -1   1    F
BITPIX      1    1   -1   2  -32
NAXIS       1    2   -1   2    0
COMMENT     3   45   47   0
HISTORY     0   -1   -1
            8    3   71   0
END         2   72   73   8
